packages/rrweb/src/record/mutation.ts

```diff
@@ -272,45 +172,137 @@ export default class MutationBuffer {
     const adds: addedNodeMutation[] = [];
     const addedIds = new Set<number>();
 
-    /**
-     * Sometimes child node may be pushed before its newly added
-     * parent, so we init a queue to store these nodes.
-     */
-    const addList = new DoubleLinkedList();
-    const getNextId = (n: Node): number | null => {
-      let ns: Node | null = n;
-      let nextId: number | null = IGNORED_NODE; // slimDOM: ignored
-      while (nextId === IGNORED_NODE) {
-        ns = ns && ns.nextSibling;
-        nextId = ns && this.mirror.getId(ns);
+    while (this.mapRemoves.length) {
+      this.mirror.removeNodeFromMap(this.mapRemoves.shift()!);
+    }
+
+    for (const n of this.movedSet) {
+      const parentNode = dom.parentNode(n);
+      if (
+        parentNode && // can't be removed if it doesn't exist
+        this.removesSubTreeCache.has(parentNode) &&
+        !this.movedSet.has(parentNode)
+      ) {
+        continue;
       }
-      return nextId;
-    };
-    const pushAdd = (n: Node) => {
-      const parent = dom.parentNode(n);
-      if (!parent || !inDom(n)) {
-        return;
+      this.addedSet.add(n);
+    }
+
+    let n: Node | null = null;
+    let parentNode: Node | null = null;
+    let parentId = -1;
+    let nextSibling: Node | null = null;
+    let ancestorBad = false;
+    const missingParents = new Set<Node>();
+    while (this.addedSet.size) {
+      if (n !== null && this.addedSet.has(n.previousSibling as Node)) {
+        // reuse parentNode, parentId, ancestorBad
+        nextSibling = n; // n is a good next sibling
+        n = n.previousSibling as Node;
+      } else {
+        n = this.addedSet.values().next().value as Node; // pop
+
+        // eslint-disable-next-line no-constant-condition
+        while (true) {
+          parentNode = dom.parentNode(n);
+          if (this.addedSet.has(parentNode as Node)) {
+            // start at top of added tree so as not to serialize children before their parents (parentId requirement)
```

Comment:
```suggestion
            // keep searching for the top of added tree so as not to serialize children before their parents (parentId requirement)
```

nit... "start at the top" threw me here and made me think 🙈